apps/aefate/src/aefa_fate.erl

```diff
@@ -306,60 +315,118 @@ bind_args_from_signature({ArgTypes, _RetSignature}, ES) ->
     Args = [aefa_engine_state:accumulator(ES) | Stack],
     bind_args(0, Args, ArgTypes, #{}, ES).
 
-check_arg_types([], _) -> ok;
-check_arg_types([T|Ts], [A|As]) ->
-    case check_type(T,A) of
-        true -> check_arg_types(Ts, As);
-        false -> {error, T, A}
+check_arg_types(Ts, As0) ->
+    As = lists:sublist(As0, length(Ts)),
+    case check_type({tuple, Ts}, {tuple, list_to_tuple(As)}) of
+        false -> {error, Ts, As};
+        Inst  -> {ok, Inst}
     end.
 
-check_all_types([], []) -> true;
-check_all_types([T|Ts], [A|As]) ->
-    check_type(T, A) andalso  check_all_types(Ts, As).
-
-check_same_type(_, []) -> true;
-check_same_type(T, [A|As]) ->
-    check_type(T, A) andalso check_same_type(T, As).
-
 bind_args(N, _, [], Mem, EngineState) ->
     ES1 = drop(N, EngineState),
     aefa_engine_state:set_memory(Mem, ES1);
 bind_args(N, [Arg|Args], [_Type|Types], Mem, EngineState) ->
     bind_args(N+1, Args, Types, Mem#{{arg, N} => {val, Arg}}, EngineState).
 
-%% TODO: Add types (and tests).
-check_type(any, _) -> true;
-check_type(_, any) -> true;
-check_type(integer, I) when ?IS_FATE_INTEGER(I) -> true;
-check_type(boolean, B) when ?IS_FATE_BOOLEAN(B) -> true;
-check_type(string, S) when ?IS_FATE_STRING(S) -> true;
-check_type(address, A) when ?IS_FATE_ADDRESS(A) -> true;
-check_type(contract, A) when ?IS_FATE_CONTRACT(A) -> true;
-check_type(oracle, A) when ?IS_FATE_ORACLE(A) -> true;
-check_type(name, A) when ?IS_FATE_NAME(A) -> true;
-check_type(bits, B) when ?IS_FATE_BITS(B) -> true;
-check_type({list, any}, L) when ?IS_FATE_LIST(L) ->
-    true;
-check_type({list, ET}, L) when ?IS_FATE_LIST(L) ->
-    check_same_type(ET, ?FATE_LIST_VALUE(L));
-check_type({tuple, Elements}, T) when ?IS_FATE_TUPLE(T) ->
-    check_all_types(Elements, ?FATE_TUPLE_ELEMENTS(T));
-check_type({map, Key, Value}, M) when ?IS_FATE_MAP(M) ->
-    {Ks, Vs} = lists:unzip(maps:to_list(?FATE_MAP_VALUE(M))),
-    check_same_type(Key, Ks) andalso
-    check_same_type(Value, Vs);
-check_type({variant, Types}, V) when ?IS_FATE_VARIANT(V) ->
-    ?FATE_VARIANT(Arities, Tag, Value) = V,
-    check_variant_arities(Types, Arities) andalso
-        check_type(lists:nth(Tag + 1, Types), ?FATE_TUPLE(Value));
-check_type(_T, _V) -> false.
-
-check_variant_arities([?FATE_TUPLE(Types)|Left1], [Arity|Left2]) ->
-    Arity =:= length(Types) andalso check_variant_arities(Left1, Left2);
-check_variant_arities([], []) ->
-    true;
-check_variant_arities(_, _) ->
-    false.
+check_type(T, V) ->
+    try
+        match_type(T, infer_type(V))
```

Comment:
I see call e.g. `aefa_fate:check_type(integer, Index)` hooked to VM (though not in Fortuna).

Even if this form (i.e. infer type, then match) is more readable than the old one (match, inferring only as much as needed), I wonder if this makes more difficult to price VM operations, and potentially enabling spending little (static) though causing node to spend large computational effort.

Reply:
`check_type` is not exported from the module.

Type inference and type checking have similar cost when successful. Inference will be more expensive to check a large structure against `integer` (say), but you would still have to pay for the size of the structure (and you have the same problem for checking if the type is not `integer` but some structured type)